// react-docgen/dist/utils/isReactCreateClassCall.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = isReactCreateClassCall;

var _astTypes = _interopRequireDefault(require("ast-types"));

var _isReactModuleName = _interopRequireDefault(require("./isReactModuleName"));

var _match = _interopRequireDefault(require("./match"));

var _resolveToModule = _interopRequireDefault(require("./resolveToModule"));

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const t = _astTypes.default.namedTypes;
/**
 * Returns true if the expression is a function call of the form
 * `React.createClass(...)`.
 */

function isReactCreateClassCallBuiltIn(path) {
  if (t.ExpressionStatement.check(path.node)) {
    path = path.get('expression');
  }

  if (!(0, _match.default)(path.node, {
    callee: {
      property: {
        name: 'createClass'
      }
    }
  })) {
    return false;
  }

  const module = (0, _resolveToModule.default)(path.get('callee', 'object'));
  return Boolean(module && (0, _isReactModuleName.default)(module));
}
/**
 * Returns true if the expression is a function call of the form
 * ```
 * import createReactClass from 'create-react-class';
 * createReactClass(...);
 * ```
 */


function isReactCreateClassCallModular(path) {
  if (t.ExpressionStatement.check(path.node)) {
    path = path.get('expression');
  }

  if (!(0, _match.default)(path.node, {
    type: 'CallExpression'
  })) {
    return false;
  }

  const module = (0, _resolveToModule.default)(path);
  return Boolean(module && module === 'create-react-class');
}
/**
 * Returns true if the expression is a function call of the form
 * `React.createClass(...)` or
 * ```
 * import createReactClass from 'create-react-class';
 * createReactClass(...);
 * ```
 */


function isReactCreateClassCall(path) {
  return isReactCreateClassCallBuiltIn(path) || isReactCreateClassCallModular(path);
}
